pkg/persistence/postgres.go

```diff
@@ -0,0 +1,387 @@
+	// returning from the function so the tx.Rollback() call will fail, trying to rollback a
+	// committed transaction. BUT, if an error occurs, we'll get our transaction rollback'ed, which
+	// is nice.
+	defer tx.Rollback()
+
+	var totalSize uint64 = 0
+	for _, file := range files {
+		totalSize += uint64(file.Size)
+	}
+
+	// This is a workaround for a bug: the database will not accept total_size to be zero.
+	if totalSize == 0 {
+		zap.L().Debug("Ignoring a torrent whose total size is zero.")
+		return nil
+	}
+
+	if exist, err := db.DoesTorrentExist(infoHash); exist || err != nil {
+		return err
+	}
+
+	var lastInsertId int64
+
+	err = tx.QueryRow(`
+		INSERT INTO `+db.schema+`.torrents (
+			info_hash,
+			name,
+			total_size,
+			discovered_on
+		) VALUES ($1, $2, $3, $4)
+		RETURNING id;
+	`, infoHash, name, totalSize, time.Now().Unix()).Scan(&lastInsertId)
+	if err != nil {
+		return errors.Wrap(err, "tx.QueryRow (INSERT INTO "+db.schema+".torrents)")
+	}
+
+	for _, file := range files {
+		if !utf8.ValidString(file.Path) {
+			zap.L().Warn(
+				"Ignoring a file whose path is not UTF-8 compliant.",
+				zap.Binary("path", []byte(file.Path)),
+			)
+
+			// Returning nil so deferred tx.Rollback() will be called and transaction will be canceled.
+			return nil
+		}
+
+		_, err = tx.Exec("INSERT INTO "+db.schema+".files (torrent_id, size, path) VALUES ($1, $2, $3);",
+			lastInsertId, file.Size, file.Path,
+		)
+		if err != nil {
+			return errors.Wrap(err, "tx.Exec (INSERT INTO "+db.schema+".files)")
+		}
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return errors.Wrap(err, "tx.Commit")
+	}
+
+	return nil
+}
+
+func (db *postgresDatabase) Close() error {
+	return db.conn.Close()
+}
+
+func (db *postgresDatabase) GetNumberOfTorrents() (uint, error) {
+	// Using estimated number of rows which can be queries much quicker
+	// https://www.postgresql.org/message-id/568BF820.9060101%40comarch.com
+	// https://wiki.postgresql.org/wiki/Count_estimate
+	rows, err := db.conn.Query(
+		"SELECT reltuples::BIGINT AS estimate_count FROM pg_class WHERE relname='" + db.schema + ".torrents';",
+	)
+	if err != nil {
+		return 0, err
+	}
+	defer rows.Close()
+
+	if rows.Next() != true {
+		return 0, fmt.Errorf("no rows returned from `SELECT reltuples::BIGINT AS estimate_count`")
+	}
+
+	// Returns int64: https://godoc.org/github.com/lib/pq#hdr-Data_Types
+	var n *uint
+	if err = rows.Scan(&n); err != nil {
+		return 0, err
+	}
+
+	// If the database is empty (i.e. 0 entries in 'torrents') then the query will return nil.
+	if n == nil {
+		return 0, nil
+	} else {
+		return *n, nil
+	}
+}
+
+func (db *postgresDatabase) QueryTorrents(
+	query string,
+	epoch int64,
+	orderBy OrderingCriteria,
+	ascending bool,
+	limit uint,
+	lastOrderedValue *float64,
+	lastID *uint64,
+) ([]TorrentMetadata, error) {
+	err := fmt.Errorf("QueryTorrents is not supported yet by PostgreSQL backend")
```

Comment:
It would be great to implement these. I will give it a try and you can perhaps help me review it?

Reply:
No problem. I'm not PostgreSQL DBA but have slight usage experience.